keps/sig-node/2008-forensic-container-checkpointing/README.md

```diff
@@ -332,7 +547,89 @@ the size of all memory pages in the checkpointed container. Each file in the con
+
+During checkpointing each memory page will be written to disk. Disk usage will increase by
+the size of all memory pages in the checkpointed container. Each file in the container that
+has been changed compared to the original version will also be part of the checkpoint.
+Disk usage will overall increase by the used memory of the container and the changed files.
+Checkpoint archive written to disk can optionally be compressed. The current implementation
+does not compress the checkpoint archive on disk.  The cluster administrator is
+responsible for monitoring disk usage and removing excess data.
+
+The kubelet will request a checkpoint from the underlying CRI implementation. In
+the checkpoint request the kubelet will specify the name of the checkpoint
+archive as `checkpoint-<podFullName>-<containerName>-<timestamp>.tar` and also
+request to store the checkpoint archive in the `checkpoints` directory below its
+root directory (as defined by `--root-dir`). This defaults to
+`/var/lib/kubelet/checkpoints`.
+
+To avoid running out of disk space an operator has been introduced: <https://github.com/checkpoint-restore/checkpoint-restore-operator>
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+Like any other kubelet API endpoint this will fail if the API server is not available.
+
+###### What are other known failure modes?
+
+- The creation of the checkpoint archive can fail.
+  - Detection: Possible return codes are:
+    - 401: Unauthorized
+    - 404: Not Found (if the ContainerCheckpoint feature gate is disabled)
+    - 404: Not Found (if the specified namespace, pod or container cannot be found)
+    - 500: Internal Server Error (if the CRI implementation encounter an error
+    during checkpointing (see error message for further details))
+    - 500: Internal Server Error (if the CRI implementation does not implement
+    the checkpoint CRI API (see error message for further details))
+    - Also see: https://kubernetes.io/docs/reference/node/kubelet-checkpoint-api/
+  - Mitigation: Do not checkpoint a container that cannot be checkpointed by CRIU.
+  - Diagnostics: The container engine will provide the location of log file created
+    by CRIU with more details.
+  - Testing: Tests are currently covering if checkpointing is enabled in the kubelet
+    or not as well as covering if the underlying container engine supports the
+    corresponding CRI API calls. The most common checkpointing failure is if the
```

Comment:
This should probably be called out more in checkpointing KEP.

What is the expected behavior if you try to checkpoint a container that is using a device?

Do we support forensic checkpoints in this case?

Reply:
Added more details about hardware specific failures.